interp/interp_test.go

```diff
@@ -191,3 +190,341 @@ func TestPiecewiseConstantPredict(t *testing.T) {
+		df func(float64) float64
+	}{
+		{
+			xs: []float64{-1.001, 0.2, 2},
+			f:  func(x float64) float64 { return x * x },
+			df: func(x float64) float64 { return 2 * x },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 2},
+			f:  func(x float64) float64 { return x * x },
+			df: func(x float64) float64 { return 2 * x },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 10},
+			f:  func(x float64) float64 { return 1.5*x - 1 },
+			df: func(x float64) float64 { return 1.5 },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 10},
+			f:  func(x float64) float64 { return -1 },
+			df: func(x float64) float64 { return 0 },
+		},
+		{
+			xs: []float64{-1.1, 0.2, 0.99, 2.5, 2.99},
+			f:  math.Cos,
+			df: math.Sin,
+		},
+		{
+			xs: []float64{-1.1, 0.2, 0.99, 2.5, 2.99},
+			f:  math.Exp,
+			df: math.Exp,
+		},
+		{
+			xs: []float64{-1.1, 0.2, 0.99, 2.5, 2.99},
+			f:  func(x float64) float64 { return math.Sin(x * x) },
+			df: func(x float64) float64 { return -2 * x * math.Cos(x*x) },
+		},
+	} {
+		ys := applyFunc(test.xs, test.f)
+		dydxs := applyFunc(test.xs, test.df)
+		var pc PiecewiseCubic
+		pc.fitWithDerivatives(test.xs, ys, dydxs)
+		n := len(test.xs)
+		for j := 0; j < n; j++ {
+			x := test.xs[j]
+			got := pc.Predict(x)
+			want := test.f(x)
+			if math.Abs(got-want) > valueTol {
+				t.Errorf("Mismatch in interpolated value at x == %g for test case %d: got %v, want %g", x, i, got, want)
+			}
+			if j < n-1 {
+				got = pc.coeffs.At(j, 0)
+				if math.Abs(got-want) > valueTol {
+					t.Errorf("Mismatch in 0-th order interpolation coefficient in %d-th node for test case %d: got %v, want %g", j, i, got, want)
+				}
+			} else {
+				got = pc.lastY
+				if math.Abs(got-want) > valueTol {
+					t.Errorf("Mismatch in lastY for test case %d: got %v, want %g", i, got, want)
+				}
+			}
+
+			if j > 0 {
+				dx := test.xs[j] - test.xs[j-1]
+				got = ((pc.coeffs.At(j-1, 3)*dx+pc.coeffs.At(j-1, 2))*dx+pc.coeffs.At(j-1, 1))*dx + pc.coeffs.At(j-1, 0)
+				if math.Abs(got-want) > valueTol {
+					t.Errorf("Interpolation coefficients in %d-th node produce mismatch in interpolated value at %g for test case %d: got %v, want %g", j-1, x, i, got, want)
+				}
+			}
+			if j == 0 {
+				got = (pc.Predict(x+h) - pc.Predict(x)) / h
+			} else if j == n-1 {
+				got = (pc.Predict(x) - pc.Predict(x-h)) / h
+			} else {
+				got = (pc.Predict(x+h) - pc.Predict(x-h)) / (2 * h)
+			}
+			want = test.df(x)
+			if math.Abs(got-want) > derivTol {
+				t.Errorf("Mismatch in interpolated derivative value at x == %g for test case %d: got %v, want %g", x, i, got, want)
+			}
+			if j < n-1 {
+				got = pc.coeffs.At(j, 1)
+				if math.Abs(got-want) > valueTol {
+					t.Errorf("Mismatch in 1-st order interpolation coefficient in %d-th node for test case %d: got %v, want %g", j, i, got, want)
+				}
+			}
+			if j > 0 {
+				dx := test.xs[j] - test.xs[j-1]
+				got = (3*pc.coeffs.At(j-1, 3)*dx+2*pc.coeffs.At(j-1, 2))*dx + pc.coeffs.At(j-1, 1)
+				if math.Abs(got-want) > valueTol {
```

Comment:
You will need to increase the tolerance here; arm64 gives a slightly different result.

```
--- FAIL: TestPiecewiseCubic (0.00s)
    interp_test.go:294: Interpolation coefficients in 3-th node produce mismatch in interpolated derivative value at 2.99 for test case 5: got 19.88568249156475, want 19.88568249156473
```

Evicting the FMAs doesn't fix this.

Reply:
Oops. Thanks for spotting.